### ISSUE TITLE
Word wraps on last letter in word

### ISSUE THREAD

**Issue by brunobowden** [WorldHealthOrganization/app#911]
With some of the recent layouts changes. The last letter of several words are being wrapped to a new line. This looks egregiously bad. We need to figure out how we better manage word wrap in the UI.

Device: iPhone 11 Max Pro
OS: 13.3.1
Settings: Accessibility off and Test Size as default

**Main screen**:
It appears that "Protect" is scaled to be correct and then "Yourself" is wrapped as it's slightly longer. The text should be scaled so that neither word is wrapped.
*"Protect
Yoursel
f"*

Similar in that "Answers" appears to be ok but "Questions" gets wrapped instead. This should appear over two lines with no word wrap.
*"Question
s &
Answers"*

**Travel Advice**
Bottom of page, it doesn't look good at all....
"General
Recommendation
s"

@hspinks - please assign to whomever makes sense
FYI @britannio @tomgilder @advayDev1

- [x] Searched the [existing issues](https://github.com/WorldHealthOrganization/app/issues) to ensure you are not creating a duplicate.
- [x] Followed the [Contributor Guidelines](https://github.com/WorldHealthOrganization/app/blob/master/docs/CONTRIBUTING.md).

![IMG_3059](https://user-images.githubusercontent.com/679339/78580598-78705580-77e7-11ea-984e-a2ceb80f29e2.PNG)
![IMG_3061](https://user-images.githubusercontent.com/679339/78580568-6d1d2a00-77e7-11ea-9196-9a96292b42fe.PNG)

**Comment by hspinks** [comment]
@brunobowden can you let us know your device type, and whether you have any special settings turned on on your phone (e.g. accessibility) to change the font size? If you could share screenshots of what you're seeing here too, that would be helpful

**Comment by brunobowden** [comment]
@hspinks - I've updated the bug with more detail. I don't believe that I have any accessibility or text size changes enabled

**Comment by hspinks** [comment]
Thanks @brunobowden! We'll take a look

**Comment by britannio** [comment]
Slightly unrelated but the IOS font is much thicker than the Android font for some reason so it'd be nice if we could use a custom font for both platforms.

**Comment by patniemeyer** [comment]
@brunobowden Can you check your Settings->Display & Brightness -> Text Size?

If we're still seeing this after it has been tweaked for small screen sizes and large adaptive text sizes then we may have to introduce the auto_size_text package.  I had been thinking it would be less disruptive to manually adjust the sizes for the first release but if we can't reproduce the problem easily we may have to do this now.

**Comment by britannio** [comment]
https://support.apple.com/en-us/HT202828
Might also be set in Settings > Accessibility, then select Display & Text Size.

**Comment by advayDev1** [comment]
Hyphenation does not work in flutter right now - https://github.com/flutter/flutter/issues/18443

There is some random pkg for it https://pub.dev/packages/hyphenator that is very immature atm

**Comment by hspinks** [comment]
Unassigning myself for someone else to pick up - Pat's suggestion in Slack to have an accessibility expert help us shape out our strategy for how to support extreme accessibility settings I think is a good one. @avkvirtru - I know you're crunching on other mocks right now, but bringing this to your attention when you have a minute to weigh in

**Comment by avkvirtru** [comment]
@hspinks I don't have anything to add atm, only that this problem may reappear with longer localized strings. For readability's sake, I would avoid breaking words in the main nav, even if hyphenation is supported. Consider wrapping the whole word that doesn't fit to the next line, smaller text, or shorter labels.

**Comment by brunobowden** [comment]
punting this for v1, we should consider this as part of the accessibility / localization update soon after v1

**Comment by tomgilder** [comment]
@advayDev1 @brunobowden PR #992 has made this happen by default on my iPhone X, and it looks terrible. Should be a P0 to fix.

![IMG_2194](https://user-images.githubusercontent.com/756862/79045156-51b08580-7c12-11ea-840f-ee9e26a5a852.PNG)

**Comment by britannio** [comment]
> @advayDev1 @brunobowden PR #992 has made this happen by default on my iPhone X, and it looks terrible. Should be a P0 to fix.
> 
> ![IMG_2194](https://user-images.githubusercontent.com/756862/79045156-51b08580-7c12-11ea-840f-ee9e26a5a852.PNG)

That was my concern @davidkaneda

**Comment by britannio** [comment]
We can tweak padding and text sizes etc but it'll still leave edge cases such as different localisations and system font sizes. 

We could check the locale, device size, and system font size and decide whether we should keep the current layout or use the following one:
![image](https://user-images.githubusercontent.com/33752528/79045443-0cd12280-7c03-11ea-961e-4ac7e94c87e0.png)

Or maybe this layout needs redesigning all together. I'm on standby to implement any necessary changes.

**Comment by kassim** [comment]
could https://pub.dev/packages/auto_size_text help?

<img src="https://raw.githubusercontent.com/leisim/auto_size_text/master/.github/art/maxlines.gif"/>

**Comment by britannio** [comment]
> could https://pub.dev/packages/auto_size_text help?
> 
> ![](https://raw.githubusercontent.com/leisim/auto_size_text/master/.github/art/maxlines.gif)

Yes but it's not ideal as it needs to be placed in a widget with a fixed height which prevents us from letting the buttons grow to fit the text(To prevent overflows). Also, it effectively caps the text size which 1. wouldn't be respecting the system's accessibility settings and 2. would only come into effect on some buttons thus making the text sizes look ununiform.

**Comment by patniemeyer** [comment]
@britannio We have a lot of control with the auto_text_size: re: 1) We can apply the system scaling to the max font size that we tell it to use ourselves and 2) auto_text_size supports the notion of "groups" that allow it to pick a uniform font size for all of the items.  Additionally @clementmouchet and I had discussed the possibility of allowing the translators to add "hints" in the ARB files for e.g. max number of lines.   I think we should set up a test harness project with a clone of our home page that lets us swap in different string lengths and scaling factors see how it behaves... then we will get a better idea of what controls are necessary.   If anyone wants to do this please ping me else I'll try to do it this weekend.

**Comment by britannio** [comment]
>We can apply the system scaling to the max font size that we tell it to use ourselves

@patniemeyer Either the maximum font size is smaller than it would've been without scaling it down or we get the issue Bruno is facing right? Maybe that's a reasonable sacrifice and I'm blowing accessibility out of proportion?

I can set up the test harness project you described.

**Comment by patniemeyer** [comment]
@britannio So what I'm envisioning is a shallow copy of our home page with two controls at the top:  1) a selector that lets us change between, say, three different sets of text (corresponding to different languages but doesn't matter here can just be English of different word length and word counts and 2) a slider that emulates the system text size setting.  Then we can drop in auto_size_text and play around with this to see what other "hints" might be necessary to get the effect that we want.  3) As a bonus it would be cool if hitting the button would pop up a dialog that lets you manually enter the text you want for the button.  This harness could later turn into a "mode" or special build of the app that the translators could use to suggest hints for the layout.

**Comment by britannio** [comment]
> @britannio So what I'm envisioning is a shallow copy of our home page with two controls at the top: 1) a selector that lets us change between, say, three different sets of text (corresponding to different languages but doesn't matter here can just be English of different word length and word counts and 2) a slider that emulates the system text size setting. Then we can drop in auto_size_text and play around with this to see what other "hints" might be necessary to get the effect that we want. 3) As a bonus it would be cool if hitting the button would pop up a dialog that lets you manually enter the text you want for the button. This harness could later turn into a "mode" or special build of the app that the translators could use to suggest hints for the layout.

On it.

**Comment by davidkaneda** [comment]
Please also pull 4px from the left/right padding of the tiles, though the
design is fragile regardless-

**Comment by advayDev1** [comment]
QA, please verify if this remains an issue, if not plz close.

**Comment by advayDev1** [comment]
Daniel's still seeing this. @britannio plz take a look at what's up with our auto-sizing if you can.

**Comment by britannio** [comment]
> Daniel's still seeing this. @britannio plz take a look at what's up with our auto-sizing if you can.

https://github.com/WorldHealthOrganization/app/pull/1007 uses fixed button heights + auto-sized text, it'd be nice to know if that works better for Daniel. I'm thinking that capping the font scale provided by the system might reduce the likelihood of this issue occurring.

**Comment by britannio** [comment]
I believe we're not using this layout anymore
